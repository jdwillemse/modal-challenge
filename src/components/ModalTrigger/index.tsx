export { default as ModalTrigger } from "./ModalTrigger";
