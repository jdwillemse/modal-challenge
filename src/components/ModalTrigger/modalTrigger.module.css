.button:not(:first-child) {
  margin-left: 0.25rem;
}

.button:not(:last-child) {
  margin-right: 0.25rem;
}
