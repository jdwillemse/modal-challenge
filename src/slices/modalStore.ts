import { create } from "zustand";

export enum ModalIDs {
  MODAL_A = "modalA",
  MODAL_B = "modalB",
}

interface ModalStore {
  modalID: null | ModalIDs;
  openModal: (id: ModalIDs) => void;
  closeModal: () => void;
}

export const useModalStore = create<ModalStore>()((set) => ({
  modalID: null,
  openModal: (id) => set({ modalID: id }),
  closeModal: () => set({ modalID: null }),
}));
