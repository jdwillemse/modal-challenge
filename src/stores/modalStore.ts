import { create } from "zustand";

export enum ModalIDs {
  ModalA = "MODAL_A",
  ModalB = "MODAL_B",
}

interface ModalStore {
  activeModalID: null | ModalIDs;
  openModal: (id: ModalIDs) => void;
  closeModal: () => void;
}

export const useModalStore = create<ModalStore>()((set) => ({
  activeModalID: null,
  openModal: (id) => set({ activeModalID: id }),
  closeModal: () => set({ activeModalID: null }),
}));
