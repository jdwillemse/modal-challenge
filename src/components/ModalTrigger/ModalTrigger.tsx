import { ReactElement, MouseEvent, useCallback } from "react";

import css from "./modalTrigger.module.css";
import { useModalStore, ModalIDs } from "../../stores/modalStore";

interface ModalTriggerType {
  id: ModalIDs;
  children: string | ReactElement;
}

function ModalTrigger({ id, children }: ModalTriggerType) {
  const { openModal } = useModalStore.getState();
  const handleModalTrigger = useCallback(
    (event: MouseEvent<HTMLButtonElement>) => {
      const { modalId } = event.currentTarget.dataset;
      if (modalId) {
        openModal(modalId as ModalIDs);
      }
    },
    [openModal]
  );

  return (
    <button
      onClick={handleModalTrigger}
      data-modal-id={id}
      type="button"
      className={css.button}
    >
      {children}
    </button>
  );
}

export default ModalTrigger;
