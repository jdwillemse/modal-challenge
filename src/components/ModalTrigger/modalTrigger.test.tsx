import { describe, expect, test } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import ModalTrigger from "./ModalTrigger";
import { useModalStore, ModalIDs } from "@stores/modalStore";

describe("modalTrigger", () => {
  const user = userEvent.setup();

  function renderModalTrigger() {
    render(<ModalTrigger id={ModalIDs.ModalA}>show modal</ModalTrigger>, {
      container: document.body,
    });
  }

  test("should render correctly", () => {
    renderModalTrigger();
    expect(screen.queryByRole("button")).toMatchSnapshot();
  });

  test("should trigger modal when clicked", async () => {
    const incSpy = vi.spyOn(useModalStore.getState(), "openModal");

    renderModalTrigger();

    await user.click(
      await screen.findByRole("button", { name: /show modal/i })
    );

    expect(incSpy).toHaveBeenCalledWith(ModalIDs.ModalA);
  });
});
