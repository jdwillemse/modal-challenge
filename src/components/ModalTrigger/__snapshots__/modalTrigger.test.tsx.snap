// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`modalTrigger > should render correctly 1`] = `
<button
  class="button"
  data-modal-id="MODAL_A"
  type="button"
>
  show modal
</button>
`;
